### ISSUE TITLE
Configure/Makefile: Perl error in FIPSMODULENAME declaration

### ISSUE THREAD

**Issue by mspncp** [openssl/openssl#13679]
_Not sure whether this classifies as an OpenSSL (documentation) issue or just a question about my Perl installation. But since others may encounter the same problem, I 'll just post it as a bug report here._

After configuring (`./Configure --debug`)  my local copy (which is at commit 52c6c12c1cad6f1046b34f4139d1aa3e967a5530), my Makefile contains the following erroneous line:

```
FIPSMODULENAME=Program fragment delivered error ``Can't locate object method "basename" via package "platform" at Configurations/unix-Makefile.tmpl line 96.''
```

which was generated from

https://github.com/openssl/openssl/blob/52c6c12c1cad6f1046b34f4139d1aa3e967a5530/Configurations/unix-Makefile.tmpl#L90-L96

Obviously, the `basename` function is missing somehow. What am I missing?

**My Perl version**

```
msp@msppc:~/src/openssl$ perl --version

This is perl 5, version 30, subversion 3 (v5.30.3) built for x86_64-linux
(with 30 registered patches, see perl -V for more detail)

Copyright 1987-2020, Larry Wall

Perl may be copied only under the terms of either the Artistic License or the
GNU General Public License, which may be found in the Perl 5 source kit.

Complete documentation for Perl, including FAQ lists, should be found on
this system using "man perl" or "perldoc perl".  If you have access to the
Internet, point your browser at http://www.perl.org/, the Perl Home Page.
```

**My Text::Template version (via Gentoo Linux Package Manager)**

```
msp@msppc:~/src/openssl$ emerge -pv Text-Template

These are the packages that would be merged, in order:

Calculating dependencies... done!
[ebuild   R    ] dev-perl/Text-Template-1.530.0::gentoo  USE="-test" 0 KiB

Total: 1 package (1 reinstall), Size of downloads: 0 KiB
```

**Comment by levitte** [comment]
There should be a `use File::basename;` in that perl nugget or an earlier one.  The obvious fix is to simply add it there

**Comment by mspncp** [comment]
But if it's just missing in the template why don't you get the error?

**Comment by mspncp** [comment]
I see a `use File::Basename` (with a capital F further down):

https://github.com/openssl/openssl/blob/52c6c12c1cad6f1046b34f4139d1aa3e967a5530/Configurations/unix-Makefile.tmpl#L1246

**Comment by levitte** [comment]
> But if it's just missing in the template why don't you get the error?

Dunno.  Maybe I have without noticing it? ....  Ah, yeah, I can see it.  I just haven't tried installing for a while, so it went unnoticed.

Anyway, the perl nuggets follow lexical order, so a `use File::basename;` *later* does no good.  Move it up or duplicate it, whichever you feel is best (perl doesn't care about duplicates).

**Comment by mspncp** [comment]
Hmmm... I tried this, first with a capital `File::Basename`, then with `File::basename`


```diff
diff --git a/Configurations/unix-Makefile.tmpl b/Configurations/unix-Makefile.tmpl
index 342e46d24d..8d71c12e44 100644
--- a/Configurations/unix-Makefile.tmpl
+++ b/Configurations/unix-Makefile.tmpl
@@ -87,7 +87,8 @@ MODULES={- join(" \\\n" . ' ' x 8,
                 fill_lines(" ", $COLUMNS - 8,
                            map { platform->dso($_) }
                            @{$unified_info{modules}})) -}
-FIPSMODULENAME={- # We do some extra checking here, as there should be only one
+FIPSMODULENAME={- use File::basename;
+                                 # We do some extra checking here, as there should be only one
                   my @fipsmodules =
                       grep { !$unified_info{attributes}->{modules}->{$_}->{noinst}
                              && $unified_info{attributes}->{modules}->{$_}->{fips} }
```


and both attempts failed:

```

FIPSMODULENAME=Program fragment delivered error ``Can't locate object method "basename" via package "platform" at Configurations/unix-Makefile.tmpl line 97.''
```

```
FIPSMODULENAME=Program fragment delivered error ``Can't locate File/basename.pm in @INC (you may need to install the File::basename module) (@INC contains: . /home/msp/src/openssl/Configurations /home/msp/src/openssl/util/perl /home/msp/perl5/lib/perl5/5.30.3/x86_64-linux /home/msp/perl5/lib/perl5/5.30.3 /home/msp/perl5/lib/perl5/x86_64-linux /home/msp/perl5/lib/perl5 /etc/perl /usr/local/lib64/perl5/5.30.3/x86_64-linux /usr/local/lib64/perl5/5.30.3 /usr/lib64/perl5/vendor_perl/5.30.3/x86_64-linux /usr/lib64/perl5/vendor_perl/5.30.3 /usr/local/lib64/perl5 /usr/lib64/perl5/vendor_perl/5.30.1/x86_64-linux /usr/lib64/perl5/vendor_perl/5.30.1 /usr/lib64/perl5/vendor_perl/5.24.3 /usr/lib64/perl5/vendor_perl/5.24.1 /usr/lib64/perl5/vendor_perl/5.22.3 /usr/lib64/perl5/vendor_perl/5.22.2 /usr/lib64/perl5/vendor_perl/5.20.2 /usr/lib64/perl5/vendor_perl/5.20.1 /usr/lib64/perl5/vendor_perl/5.18.2 /usr/lib64/perl5/vendor_perl/5.16.3 /usr/lib64/perl5/vendor_perl /usr/lib64/perl5/5.30.3/x86_64-linux /usr/lib64/perl5/5.30.3 /home/msp/perl5/lib/perl5/5.30.1/x86_64-linux /home/msp/perl5/lib/perl5/5.30.1 /home/msp/perl5/lib/perl5/5.24.3 /home/msp/perl5/lib/perl5/5.24.1 /home/msp/perl5/lib/perl5/5.22.3 /home/msp/perl5/lib/perl5/5.22.2 /home/msp/perl5/lib/perl5/5.20.2 /home/msp/perl5/lib/perl5/5.20.1 /home/msp/perl5/lib/perl5/5.18.2 /home/msp/perl5/lib/perl5/5.16.3 /home/msp/src/openssl/external/perl/Text-Template-1.56/lib) at Configurations/unix-Makefile.tmpl line 90.
BEGIN failed--compilation aborted at Configurations/unix-Makefile.tmpl line 90.''
```


I guess, I'll better just wait till you come up with a PR ;-)

**Comment by mspncp** [comment]
BTW: wouldn't it be better it `./Configure` would break on this kind of errors, instead of writing garbage into the makefile`?

**Comment by levitte** [comment]
If should be File::Basename (capital B), and is a core perl 5 module since the beginning of time (well, since perl 5.3.7, so very early).  I have *no* idea how gentoo handles perl modules...  I know that in RPM based Linuxes, you have to install `perl-core` rather than `perl` (the latter is *very* bare bones), so that's beyond what I can do in a PR.

> BTW: wouldn't it be better it `./Configure` would break on this kind of errors, instead of writing garbage into the makefile?

I agree.  Strictly speaking, it's util/dofile.pl that does the work, and while it does catch some errors, there are errors, like this one, that escapes the mechanism, and I haven't figured out why.

**Comment by mspncp** [comment]
> If should be File::Basename (capital B), and is a core perl 5 module

The module exists and I can load without problem, if I do it standalone.

```
use File::Basename;

print basename "/foo/bar/baz"
```
I rather suspect that the problem has something to do with the
```
join(" ", map { basename **platform**->dso($_) } @fipsmodules)

```
snippet, because the error message talks about loading "basename" via the **"platform"** package.

```
``Can't locate object method "basename" via package "platform" at Configurations/unix-Makefile.tmpl line 97.''
```

```
FIPSMODULENAME={- use File::Basename;
                  # We do some extra checking here, as there should be only one
                  my @fipsmodules =
                      grep { !$unified_info{attributes}->{modules}->{$_}->{noinst}
                             && $unified_info{attributes}->{modules}->{$_}->{fips} }
                      @{$unified_info{modules}};
                  die "More that one FIPS module" if scalar @fipsmodules > 1;
                  join(" ", map { basename platform->dso($_) } @fipsmodules) -}
```

**Comment by levitte** [comment]
This is one of those where perl can be...  mysterious, even for me.  This has to do with how it parses statements in certain blocks (`map` blocks, for example), which can sometimes seem heuristic more than anything else.

If you change `basename platform->dso($_)` to `basename(platform->dso($_))`, you'll probably see a better result.

**Comment by mattcaswell** [comment]
This was fixed by #14320
